aggregator/entry.go

```diff
@@ -315,170 +339,179 @@ func (e *Entry) recordLastAccessed(currTime time.Time) {
-	}
-	// If this is a stale policy, we don't update the existing policy
-	// and instead use the cached policy.
-	if sp.CutoverNanos < e.cutoverNanos {
-		e.metrics.stalePolicy.Inc(1)
-		return false
-	}
-	if e.cutoverNanos != sp.CutoverNanos {
-		return true
-	}
-	policies, useDefaultPolicies := sp.Policies()
-	if e.useDefaultPolicies && useDefaultPolicies {
+// NB: The metadata passed in is guaranteed to have cut over based on the current time.
+func (e *Entry) shouldUpdateMetadatasWithLock(
+	currTime time.Time,
+	sm metadata.StagedMetadata,
+) bool {
+	// If this is a stale metadata, we don't update the existing metadata.
+	if e.cutoverNanos > sm.CutoverNanos {
+		e.metrics.staleMetadata.Inc(1)
 		return false
 	}
-	if useDefaultPolicies {
-		policies = e.opts.DefaultPolicies()
-	}
-	return e.hasPolicyChangesWithLock(policies)
-}
 
-func (e *Entry) hasPolicyChangesWithLock(newPolicies []policy.Policy) bool {
-	if len(e.aggregations) != len(newPolicies) {
+	// If this is a newer metadata, we always update.
+	if e.cutoverNanos < sm.CutoverNanos {
 		return true
 	}
-	for _, policy := range newPolicies {
-		if _, exists := e.aggregations[policy]; !exists {
-			return true
+
+	// Iterate over the list of pipelines and check whether we have metadata changes.
+	// NB: If the incoming metadata have the same set of aggregation keys as the cached
+	// set but also have duplicates, there is no need to update metadatas as long as
+	// the cached set has all aggregation keys in the incoming metadata and vice versa.
+	bs := bitset.New(uint(len(e.aggregations)))
+	for _, pipeline := range sm.Pipelines {
+		storagePolicies := e.storagePolicies(pipeline.StoragePolicies)
+		for _, storagePolicy := range storagePolicies {
+			key := aggregationKey{
+				aggregationID: pipeline.AggregationID,
+				storagePolicy: storagePolicy,
+				pipeline:      pipeline.Pipeline,
+			}
+			idx := e.aggregations.index(key)
+			if idx < 0 {
+				return true
+			}
+			bs.Set(uint(idx))
 		}
 	}
-	return false
+	return bs.All(uint(len(e.aggregations)))
 }
 
-func (e *Entry) updatePoliciesWithLock(
-	typ unaggregated.Type,
-	id metricid.RawID,
-	ownsID bool,
-	hasDefaultPoliciesList bool,
-	sp policy.StagedPolicies,
-) error {
-	policies, useDefaultPolicies := sp.Policies()
-	if useDefaultPolicies {
-		policies = e.opts.DefaultPolicies()
+func (e *Entry) storagePolicies(policies []policy.StoragePolicy) []policy.StoragePolicy {
+	if !policy.IsDefaultStoragePolicies(policies) {
+		return policies
 	}
+	return e.opts.DefaultStoragePolicies()
+}
 
-	// Fast path to exit in case the policies didn't change.
-	if !e.hasPolicyChangesWithLock(policies) {
-		e.hasDefaultPoliciesList = hasDefaultPoliciesList
-		e.useDefaultPolicies = useDefaultPolicies
-		e.cutoverNanos = sp.CutoverNanos
-		e.metrics.policyUpdates.Inc(1)
-		return nil
+func (e *Entry) maybeCopyIDWithLock(metric unaggregated.MetricUnion) metricid.RawID {
+	// If we own the ID, there is no need to copy.
+	if metric.OwnsID {
```

Comment:
when does the metric not own its ID?

Reply:
So right now the iterator always reuses the ID during decoding, because the iterator is per-connection, it can reuse ID across iterations as long as the ID is copied after the iterator moves onto the next iteration.